### PR TITLE
RDS: Generate random password and store in referenced secret

### LINKED ISSUE #667: DBCluster spec is different from RDSInstance on requirement of `masterUserPasswordRef` field
### What problem are you facing?

The `DBCluster` resource has `masterUserPasswordRef` marked as a required field, whereas the `RDSInstance` resource has it marked as optional. 

The `RDSInstance` controller will generate a random password if `masterUserPasswordRef` is not provided, but the `DBCluster` controller is autogenerated and does not. 

### How could Crossplane help solve your problem?

Provide guidance on how to add hooks to autogenerated resources so I can contribute back a password creation workflow similar to how the `RDSInstance` controller does things. I'm willing to put in the development time to make the necessary changes.

### PR DESCRIPTION
**Pull request by PocketMobsters**
### Description of your changes

Addresses these use cases:

1. DBCluster API has master password as optional
    a. `masterUserPasswordSecretRef` is now optional, controller ignores not found errors and returns empty string for password
2. DBCluster does not autogenerate a password
    a. Added `autogeneratePassword` which, if `masterUserPasswordSecretRef` is set, will create a random password and store it in the specified location.

Fixes #667 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Generated new CRDs and applied to a cluster that was *not* running provider-aws. Ran provider-aws locally and verified that:

1. masterUserPasswordSecretRef being optional is respected and will create a DBCluster if not provided.
2. If masterUserPasswordSecretRef is set, it will create a DBCluster with the specified password
3. If autogeneratePassword and masterUserPasswordSecretRef is set, and *no* input password is located, it will generate a random password and store at the location specified in masterUserPasswordSecretRef
4. If autogeneratePassword and masterUserPasswordSecretRef is set, and an input password is located, it will be used instead of a random password  

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by PocketMobsters** [comment]
@muvaf I'm willing to do similar work on the RDSInstance controller if this is an acceptable pattern, granted RDSInstance doesn't use hooks so it'll be ever so slightly different

**Comment by PocketMobsters** [comment]
Dropped the RDSInstance changes. As per the discussion in Slack the path forward is to not make any changes to the RDSInstance resource contract regarding password management, and to have that work be done on a future resource `DBInstance` instead.

File: pkg/controller/rds/dbcluster/setup.go [hunk 1]
```diff
@@ -81,10 +91,19 @@ type custom struct {
 }
 
 func (e *custom) preCreate(ctx context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.CreateDBClusterInput) error {
-	pw, _, err := rds.GetPassword(ctx, e.kube, &cr.Spec.ForProvider.MasterUserPasswordSecretRef, cr.Spec.WriteConnectionSecretToReference)
-	if err != nil {
+	pw, _, err := rds.GetPassword(ctx, e.kube, cr.Spec.ForProvider.MasterUserPasswordSecretRef, cr.Spec.WriteConnectionSecretToReference)
+	if resource.IgnoreNotFound(err) != nil {
 		return errors.Wrap(err, "cannot get password from the given secret")
 	}
+	if pw == "" && cr.Spec.ForProvider.AutogeneratePassword != nil && *cr.Spec.ForProvider.AutogeneratePassword {
```
Comment (muvaf) [inline pkg/controller/rds/dbcluster/setup.go:98]:
```suggestion
	if pw == "" && awsclients.BoolValue(cr.Spec.ForProvider.AutogeneratePassword) {
```

File: apis/rds/v1alpha1/custom_types.go [hunk 2]
```diff
@@ -49,8 +58,9 @@ type CustomDBClusterParameters struct {
 	// The password for the master database user. This password can contain any
 	// printable ASCII character except "/", """, or "@".
 	//
-	// Constraints: Must contain from 8 to 41 characters. Required.
-	MasterUserPasswordSecretRef xpv1.SecretKeySelector `json:"masterUserPasswordSecretRef"`
+	// Constraints: Must contain from 8 to 41 characters.
+	// +optional
+	MasterUserPasswordSecretRef *xpv1.SecretKeySelector `json:"masterUserPasswordSecretRef"`
```
Comment (muvaf) [inline apis/rds/v1alpha1/custom_types.go:63]: Does AWS API allow creating `DBCluster` without a password?

File: pkg/controller/rds/dbcluster/setup.go [hunk 3]
```diff
@@ -131,3 +152,42 @@ func filterList(cr *svcapitypes.DBCluster, obj *svcsdk.DescribeDBClustersOutput)
...
+	nn := types.NamespacedName{
+		Name:      ref.Name,
+		Namespace: ref.Namespace,
+	}
+	sc := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nn.Name,
+			Namespace: nn.Namespace,
+		},
+	}
```
Comment (muvaf) [inline pkg/controller/rds/dbcluster/setup.go:170]:
```suggestion
	sc := &corev1.Secret{}
```

`Get` fills that for you.

File: pkg/controller/rds/dbcluster/setup.go [hunk 4]
```diff
@@ -131,3 +152,42 @@ func filterList(cr *svcapitypes.DBCluster, obj *svcsdk.DescribeDBClustersOutput)
...
+	err := e.kube.Get(ctx, nn, sc)
+	var create bool
```
Comment (muvaf) [inline pkg/controller/rds/dbcluster/setup.go:172]: Can we use `APIPatchingApplicator` or `APIUpdatingApplicator` from crossplane runtime?

File: apis/rds/v1alpha1/custom_types.go (hunk 2 above)
Comment (PocketMobsters) [reply apis/rds/v1alpha1/custom_types.go:63]: Yeah, it's optional.

**Comment by riveryc** [comment]
@PocketMobsters Thanks for your contribution. looking forward to see it is merged.

**Comment by haarchri** [comment]
@PocketMobsters Happy new year any chance to get the Open Points resolved ?

**Comment by haarchri** [comment]
@PocketMobsters any chance to see this finalized ?

**Comment by haarchri** [comment]
Implemented with #1413  @PocketMobsters thanks for working on this - some Folks implemented your Branch with up2date master

**Pull request closed without merging**